### DISCUSSION TITLE
What options in policy.xml should I change?

### DISCUSSION THREAD

**Discussion by igormironchik** [ImageMagick/ImageMagick#6733] [Help] [answered]
### ImageMagick version

6.9.11-60

### Operating system, version and so on

KDE neon 5.27

### Description

I have the following `policy.xml` file for `ImageMagick 6`.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE policymap [
  <!ELEMENT policymap (policy)*>
  <!ATTLIST policymap xmlns CDATA #FIXED ''>
  <!ELEMENT policy EMPTY>
  <!ATTLIST policy xmlns CDATA #FIXED '' domain NMTOKEN #REQUIRED
    name NMTOKEN #IMPLIED pattern CDATA #IMPLIED rights NMTOKEN #IMPLIED
    stealth NMTOKEN #IMPLIED value CDATA #IMPLIED>
]>
<policymap>
  <policy domain="resource" name="temporary-path" value="/tmp"/>
  <policy domain="resource" name="memory" value="3GiB"/>
  <policy domain="resource" name="map" value="512MiB"/>
  <policy domain="resource" name="width" value="8KP"/>
  <policy domain="resource" name="height" value="8KP"/>
  <policy domain="resource" name="list-length" value="128"/>
  <policy domain="resource" name="area" value="5MP"/>
  <policy domain="resource" name="disk" value="50GiB"/>
  <policy domain="delegate" rights="none" pattern="URL" />
  <policy domain="delegate" rights="none" pattern="HTTPS" />
  <policy domain="delegate" rights="none" pattern="HTTP" />
  <policy domain="path" rights="none" pattern="@*"/>
  <policy domain="coder" rights="none" pattern="PS" />
  <policy domain="coder" rights="none" pattern="PS2" />
  <policy domain="coder" rights="none" pattern="PS3" />
  <policy domain="coder" rights="none" pattern="EPS" />
  <policy domain="coder" rights="none" pattern="PDF" />
  <policy domain="coder" rights="none" pattern="XPS" />
</policymap>
```

When I work with `ImageMagick` it uses `3GiB` of memory and `512MiB` of shared memory, as I requested. When memory exceeded `ImageMagick` begins to write files on disk in `/tmp` folder. OK, everything is fine, but it writes files with `~7MiB` size and `~500` files only. It uses on disk `~3.5GiB` only.

So my questions are:

* What does condition a size of a temporary file? (`~7MiB`)
* Why `ImageMagick` doesn't use all `50GiB` of allowed disk space?

**Comment by snibgo** [comment]
Those limits are per command. IM won't use more resources than it needs. For example, a simple task might need only 1 MiB of memory. If IM runs out of main memory, it will use disk. If it runs out of disk, the command will fail.

50 GiB is a generous limit. Each pixel might take 8 bytes, so that would be about 6 billion pixels. If you have 10 images in memory at the same time, each image could have 600 million pixels. There would be 10 files, each about 4.8 GiB. If your images are smaller, or you have fewer in memory at the same time, IM will use less disk.

**Comment by igormironchik** [comment]
I see it. I need to store in memory a big amount of images (1000x1000), for example. It's 1MP per image. And I need to have possibility to use all 50 GiB of disk, and fail only after all limits exceeded. But with the above `policy.xml` I see that IM uses 3 GiB of memory, 512 MiB of shared memory and only ~ 4 GiB of disk space. Allocation further throws an exception, but why? I have ~46 GiB of disk yet.

**Comment by snibgo** [comment]
How many images do you have in memory at the same time? 10, 100, 1000?

> Allocation further throws an exception,

What is the error message? If a delegate is failing to allocate memory, you could lower your limits so IM will use less for itself. But if IM is failing to allocate memory, you may need to raise the limits.

IM also needs about 27 thousand bytes per image. In addition, delegates that read and write image files (such as libpng, which reads and writes PNG files) need memory, and these usually need real memory, not disk.

**Comment by urban-warrior** [comment]
See https://github.com/ImageMagick/ImageMagick/discussions/6555.  We now offer reasonable baseline security policies depending on the requirements of your local environment. Also, read about large file support @ https://imagemagick.org/script/architecture.php#tera-pixel.  To reduce the memory footprint, for example, adjust the limits to force the pixel cache to disk providing for more memory to the ImageMagick delegates which may not have optimized resource usage.

**Comment by igormironchik** [comment, marked as answer]
> What is the error message? If a delegate is failing to allocate memory, you could lower your limits so IM will use less for itself. But if IM is failing to allocate memory, you may need to raise the limits.

Oh, thank you for the question. It says:

> `unable to write pixel cache '/tmp/magick-Fp4qxCMsz3FD_8AMtiQFbpo2AasWtokc': No space left on device @ error/cache.c/WritePixelCachePixels/6012`

I didn't know that `/tmp` is limitetd to 4 GiB on my system...
